File: cucumber-expressions/javascript/src/regular_expression.js
const Argument = require('./argument')
const TreeRegexp = require('./tree_regexp')
const ParameterType = require('./parameter_type')

class RegularExpression {
  constructor(expressionRegexp, parameterTypeRegistry) {
    this._expressionRegexp = expressionRegexp
    this._parameterTypeRegistry = parameterTypeRegistry
    this._treeRegexp = new TreeRegexp(expressionRegexp)
  }

  match(text) {
    const parameterTypes = this._treeRegexp.groupBuilder.children.map(
      groupBuilder => {
        const parameterTypeRegexp = groupBuilder.source

        return (
          this._parameterTypeRegistry.lookupByRegexp(
            parameterTypeRegexp,
            this._treeRegexp,
            text
          ) ||
          new ParameterType(
            null,
            parameterTypeRegexp,
            String,
            s => s,
            false,
            false
          )
        )
      }
    )

    return Argument.build(this._treeRegexp, text, parameterTypes)
  }

  get regexp() {
    return this._expressionRegexp
  }

  get source() {
    return this._expressionRegexp.source
  }
}

module.exports = RegularExpression

File: cucumber-expressions/go/regular_expression.go
package cucumberexpressions

import (
	"reflect"
	"regexp"
)

type RegularExpression struct {
	expressionRegexp      *regexp.Regexp
	parameterTypeRegistry *ParameterTypeRegistry
	treeRegexp            *TreeRegexp
}

func NewRegularExpression(expressionRegexp *regexp.Regexp, parameterTypeRegistry *ParameterTypeRegistry) *RegularExpression {
	return &RegularExpression{
		expressionRegexp:      expressionRegexp,
		parameterTypeRegistry: parameterTypeRegistry,
		treeRegexp:            NewTreeRegexp(expressionRegexp),
	}
}

func (r *RegularExpression) Match(text string, typeHints ...reflect.Type) ([]*Argument, error) {
	parameterTypes := []*ParameterType{}
	for i, groupBuilder := range r.treeRegexp.GroupBuilder().Children() {
		parameterTypeRegexp := groupBuilder.Source()
		typeHint := reflect.TypeOf("")
		hasTypeHint := i < len(typeHints)
		if hasTypeHint {
			typeHint = typeHints[i]
		}

		parameterType, err := r.parameterTypeRegistry.LookupByRegexp(parameterTypeRegexp, r.expressionRegexp.String(), text)
		if err != nil {
			return nil, err
		}
		if parameterType!= nil && hasTypeHint && !parameterType.UseRegexpMatchAsStrongTypeHint() {
			if parameterType.Type() != typeHint.Name() {
				parameterType = nil
			}
		}

		if parameterType == nil {
			anonymousParameterType, err := createAnonymousParameterType(parameterTypeRegexp)
			if err != nil {
				return nil, err
			}
			parameterType = anonymousParameterType
		}

		if parameterType.isAnonymous() {
			deanonimizedParameterType, err := parameterType.deAnonymize(typeHint, r.objectMapperTransformer(typeHint))
			if err != nil {
				return nil, err
			}
			parameterType = deanonimizedParameterType
		}
		parameterTypes = append(parameterTypes, parameterType)
	}
	return BuildArguments(r.treeRegexp, text, parameterTypes), nil
}

func (r *RegularExpression) Regexp() *regexp.Regexp {
	return r.expressionRegexp
}

func (r *RegularExpression) Source() string {
	return r.expressionRegexp.String()
}

func (r *RegularExpression) objectMapperTransformer(typeHint reflect.Type) func(args ...*string) interface{} {
	return func(args ...*string) interface{} {
		i, err := r.parameterTypeRegistry.defaultTransformer.Transform(*args[0], typeHint)
		if err != nil {
			panic(err)
		}
		return i
	}
}
